### ISSUE TITLE
[3.x] Load all routes for testing purposes

### ISSUE THREAD

**Issue by ravage84** [cakephp/cakephp#15086]
Feature Request

* CakePHP Version: 3.9.x (applies to all versions where plugin routes are loaded through Application.php/the middleware stack, probably applies to 4.x, too)

To test an application with a rather sizable amount of routes, I want to load all routes of the application, including the plugin routes, and inspect them.

Currently, I use ``Router::routes();`` to load the routes of the application itself (app) , but I see no conventient way to load the routes of all or (preferably) some plugins.

I think adding such a functionality, would ease such testing cases.

**Comment by markstory** [comment]
What kind of tests are you looking to be able to write?

**Comment by ravage84** [comment]
@markstory We test the amount of routes (grand total but also specific sub totals).

**Comment by othercorey** [comment]
When you add a route you just change the total number in the assert? What does that catch?

**Comment by markstory** [comment]
If you are loading plugins with the `loadPlugins()` method on TestCase, it will also load routes for those plugins. That should make them available via `Router::routes()`.

**Comment by ravage84** [comment]
@markstory many thanks! That did the trick.

I didn't know of this method.
I guess we setup this test case before [the method became available (3.7)](https://book.cakephp.org/3/en/appendices/3-7-migration-guide.html#testsuite).

I solved it like this:

```` php
    /**
     * Load local plugins
     *
     * Only loads plugins from the ``plugins`` folder.
     * Does not load third party from the ``vendor`` folder.
     *
     * @return void
     */
    public function loadLocalPlugins()
    {
        $localPlugins = glob(ROOT . '/plugins/*', GLOB_ONLYDIR);
        $localPluginNames = array_map(function ($localPluginPath) {
            return basename($localPluginPath);
        }, $localPlugins);
        $this->loadPlugins($localPluginNames);
    }

$this->loadLocalPlugins();

$routes = Router::routes();

$pluginGroupedRoutes = $routes->groupBy(function ($route) {
    return isset($route->defaults['plugin']) ? $route->defaults['plugin'] : 'app';
})->toArray();

$this->assertCount(10, $pluginGroupedRoutes['PluginA'], 'Plugin route count does not match');
$this->assertCount(5, $pluginGroupedRoutes['PluginB'], 'Plugin route count does not match');
$this->assertCount(120, $pluginGroupedRoutes['app'], 'App route count does not match');
````

Do you guys think the method ``loadLocalPlugins()`` would be useful if added to ``IntegrationTestTrait``?

And/or may be we add an example to the [testing section of the cookbook](https://book.cakephp.org/3/en/development/testing.html#integration-testing-psr-7-middleware)?

**Comment by markstory** [comment]
An example in the book would be great.
